File: mobile/targets/widget/Snapshot.swift
import Foundation

/// Mirrors `WidgetSnapshot` in mobile/src/widget/snapshot.ts.
///
/// When you add a field on the TypeScript side, add it here too — the
/// `JSONDecoder` will silently skip unknown keys (no crash), but the
/// new field won't appear in the widget UI until both sides agree on
/// the shape.
struct WidgetSnapshot: Codable {
  let updatedAt: String
  let netCashMtd: NetCash
  let cashAccounts: [CashAccount]
  let totalCash: Double

  struct NetCash: Codable {
    let income: Double
    let spending: Double
    let net: Double
    let transactionCount: Int
  }

  struct CashAccount: Codable, Identifiable {
    let id: Int
    let displayName: String
    let institutionName: String?
    let mask: String?
    let balance: Double
  }
}

extension WidgetSnapshot {
  /// Shared App Group identifier. Must match
  /// `mobile/modules/widget-bridge/ios/WidgetBridgeModule.swift`
  /// and `app.json` ios.entitlements.
  static let appGroup = "group.com.tuskledger.mobile"
  static let snapshotKey = "snapshot"

  /// Read the latest snapshot from the shared App Group container.
  /// Returns nil if the app hasn't run a sync yet or the entitlement
  /// is misconfigured (suite name returns nil).
  static func read() -> WidgetSnapshot? {
    guard
      let defaults = UserDefaults(suiteName: Self.appGroup),
      let json = defaults.string(forKey: Self.snapshotKey),
      let data = json.data(using: .utf8)
    else { return nil }
    return try? JSONDecoder().decode(WidgetSnapshot.self, from: data)
  }

  /// Used by the widget gallery and on cold-start before any sync has
  /// landed. Numbers are recognizable-but-fake so it's obviously a
  /// preview, not real data.
  static let placeholder = WidgetSnapshot(
    updatedAt: ISO8601DateFormatter().string(from: Date()),
    netCashMtd: .init(income: 5200, spending: 3400, net: 1800, transactionCount: 142),
    cashAccounts: [
      .init(id: 1, displayName: "Chase Checking",    institutionName: "Chase",       mask: "1234", balance: 8421.18),
      .init(id: 2, displayName: "Ally Savings",      institutionName: "Ally Bank",   mask: "9876", balance: 4302.07),
      .init(id: 3, displayName: "Cash Reserve",      institutionName: "Wealthfront", mask: "5544", balance: 1011.55),
    ],
    totalCash: 13734.80
  )
}

File: mobile/src/widget/snapshot.ts
/**
 * Widget snapshot — the small precomputed payload the iOS home-screen
 * widget reads on every render.
 *
 * The widget extension runs in its own process and can't reach into
 * the React Native runtime or the SQLite mirror directly. So every
 * time the phone finishes a sync, we:
 *
 *   1. compute a flat JSON snapshot from local SQLite (cheap, ~ms),
 *   2. write it to UserDefaults inside the shared App Group container
 *      via the `widget-bridge` native module, and
 *   3. tell WidgetKit to reload all timelines so the widget picks the
 *      new snapshot up immediately.
 *
 * The widget falls back to a baked-in placeholder when nothing is in
 * the container yet (fresh install, never paired) so the widget
 * gallery still has something to show.
 *
 * The shape here MUST match the `WidgetSnapshot` Codable struct on
 * the Swift side (mobile/targets/widget/Snapshot.swift). Keep them
 * in sync — adding a field on one side without the other will not
 * crash but the new field will silently fail to render.
 */
import { AccountRow, currentMonthSummary, listAccounts } from '../db/queries';
// Relative import: the native side of this local Expo module is autolinked
// from modules/, but Metro can't resolve the bare 'widget-bridge' name
// without a node_modules entry — so import the JS entry directly.
import * as WidgetBridge from '../../modules/widget-bridge/src/index';

export interface CashAccountSnapshot {
  id: number;
  displayName: string;
  institutionName: string | null;
  mask: string | null;
  balance: number;
}

export interface WidgetSnapshot {
  /** ISO timestamp of when this snapshot was generated. */
  updatedAt: string;
  netCashMtd: {
    income: number;
    spending: number;
    net: number;
    transactionCount: number;
  };
  /** Depository accounts sorted by current_balance descending. */
  cashAccounts: CashAccountSnapshot[];
  /** Sum of cashAccounts[].balance — precomputed so the widget doesn't iterate. */
  totalCash: number;
}

/**
 * "Cash" matches the dashboard's depository bucket (checking +
 * savings + money market + CD). The web `CashBalances` tile is
 * checking-only because its job is bill-bounce alerts; the widget
 * is broader because a glance at the lock screen wants total liquid
 * cash across all banking accounts, not just one subtype.
 */
function isCashAccount(a: AccountRow): boolean {
  return (a.type || '').toLowerCase() === 'depository';
}

export async function buildSnapshot(): Promise<WidgetSnapshot> {
  const [allAccounts, summary] = await Promise.all([
    listAccounts(),
    currentMonthSummary(),
  ]);

  const cashAccounts: CashAccountSnapshot[] = allAccounts
    .filter(isCashAccount)
    .map((a) => ({
      id: a.id,
      displayName: a.custom_name || a.name,
      institutionName: a.institution_name,
      mask: a.mask,
      balance: a.current_balance ?? 0,
    }))
    .sort((x, y) => y.balance - x.balance);

  const totalCash = cashAccounts.reduce((s, a) => s + a.balance, 0);

  return {
    updatedAt: new Date().toISOString(),
    netCashMtd: summary,
    cashAccounts,
    totalCash,
  };
}

/**
 * Persist the latest snapshot to the App Group container and reload
 * widget timelines. Best-effort: any failure is swallowed so the
 * sync never fails because the widget couldn't refresh.
 *
 * Common failure modes that are fine to ignore here:
 *   - Running in Expo Go (the widget-bridge native module isn't
 *     present, so requireNativeModule throws on first use).
 *   - App Group entitlement missing (only happens during initial
 *     bring-up before the first signed dev build).
 */
export async function publishSnapshot(): Promise<void> {
  try {
    const snapshot = await buildSnapshot();
    await WidgetBridge.writeSnapshot(JSON.stringify(snapshot));
    await WidgetBridge.reloadAll();
  } catch (e) {
    // Non-critical — surface to console but never to the user
    if (__DEV__) console.warn('[widget] publishSnapshot failed:', e);
  }
}

/**
 * Clear the App Group snapshot so the home-screen widget stops showing
 * real balances after an unpair or a 401 token-revoke wipe. Without
 * this the widget keeps rendering the last-published snapshot — real
 * cash totals — indefinitely, because the snapshot is only ever
 * overwritten on a *successful* sync (which no longer happens once
 * we're unpaired).
 *
 * Writing an empty string makes the Swift side fall back to its baked-in
 * placeholder (same path as a fresh, never-paired install). Best-effort:
 * any failure is swallowed for the same reasons publishSnapshot swallows.
 */
export async function clearSnapshot(): Promise<void> {
  try {
    await WidgetBridge.writeSnapshot('');
    await WidgetBridge.reloadAll();
  } catch (e) {
    if (__DEV__) console.warn('[widget] clearSnapshot failed:', e);
  }
}
